### PR TITLE
Fail fast remote cluster requests

### LINKED ISSUE #74773: CCS requests to disconnected cluster all wait for a connection timeout
**Description of the problem including expected versus actual behavior**:

Today when sending a request to a remote cluster we first ensure that the remote is connected:

https://github.com/elastic/elasticsearch/blob/a92a647b9f17d1bddf5c707490a19482c273eda3/server/src/main/java/org/elasticsearch/transport/RemoteClusterAwareClient.java#L36-L37

If we don't have an open connection then this code attempts to open one. If the remote cluster is partitioned away then this attempt will time out (after 30s I believe) at which point the response handler receives the failure. If we attempt to send multiple requests to the remote cluster in sequence then these timeouts stack up, potentially resulting in many minutes of delay for the overall request.

This behaviour is ok for CCR, but a 30s delay in a cross-cluster search seems less acceptable, especially if the remotes are configured with `skip_unavailable: true`. I think it would often be preferable to fail search requests to disconnected clusters immediately instead.

**Steps to reproduce**:

 1. Establish a remote-cluster connection between two clusters (optionally with `skip_unavailable: true`)
 2. Introduce a black-hole-style network partition between the clusters
 3. Perform a cross-cluster search and observe that each search takes 30s to fail.

### PR DESCRIPTION
**Pull request by ywelsch**
When using the skip_unavailable option on remote clusters, searches and similar actions should fail quickly instead of waiting on connections to be reestablished.

This PR makes it so that search, field caps, resolve index, and term enum cross-cluster requests fail fast when skip_unavailable is set to true on the given cluster. Internal cross-cluster requests (e.g. license checks, CCR, etc.) still continue waiting on connection attempts.

Closes #74773

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterAwareClient.java [hunk 1]
```diff
@@ -37,6 +45,16 @@
...
+            try {
+                remoteClusterService.getConnection(clusterAlias);
+            } catch (NoSuchRemoteClusterException e) {
+                listener.onFailure(e);
+                // trigger another connection attempt, but don't wait for it to complete
+                remoteClusterService.ensureConnected(clusterAlias, ActionListener.wrap(() -> {}));
+                return;
+            }
+        }
         remoteClusterService.ensureConnected(clusterAlias, ActionListener.wrap(v -> {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/RemoteClusterAwareClient.java:58]: This still risks a wait in cases where we discovered the remote cluster was disconnected just after calling `getConnection()` above. I think we should avoid waiting on the result of `ensureConnected()` at all if `this.ensureConnected` is false.

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/transport/RemoteClusterAwareClient.java:58]: If the connection is blackholed (and still appears healthy), we might also wait on that. I'm not sure if it's worth complicating the implementation here for this edge case. This is a best effort check for cases where connections are unavailable for a prolonged time, not a perfect failure detector.

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/transport/RemoteClusterAwareClient.java:58]: The trouble I see is that we can handle an active-but-blackholed connection pretty quickly (a few seconds of retries is enough) but the wait for a new connection is always much slower. With a high enough rate of searches we could end up with many of them blocked within `ensureConnected`. I was thinking of something like the following, are you sure this is too complicated?

```java
    @Override
    protected <Request extends ActionRequest, Response extends ActionResponse> void doExecute(
        ActionType<Response> action,
        Request request,
        ActionListener<Response> listener
    ) {
        maybeEnsureConnected(ActionListener.wrap(v -> {
            final Transport.Connection connection;
            try {
                if (request instanceof RemoteClusterAwareRequest) {
                    DiscoveryNode preferredTargetNode = ((RemoteClusterAwareRequest) request).getPreferredTargetNode();
                    connection = remoteClusterService.getConnection(preferredTargetNode, clusterAlias);
                } else {
                    connection = remoteClusterService.getConnection(clusterAlias);
                }
            } catch (NoSuchRemoteClusterException e) {
                if (ensureConnected == false) {
                    // trigger another connection attempt, but don't wait for it to complete
                    remoteClusterService.ensureConnected(clusterAlias, ActionListener.wrap(() -> {}));
                }
                throw e;
            }
            service.sendRequest(
                connection,
                action.name(),
                request,
                TransportRequestOptions.EMPTY,
                new ActionListenerResponseHandler<>(listener, action.getResponseReader())
            );
        }, listener::onFailure));
    }

    private void maybeEnsureConnected(ActionListener<Void> ensureConnectedListener) {
        if (ensureConnected) {
            remoteClusterService.ensureConnected(clusterAlias, ensureConnectedListener);
        } else {
            ensureConnectedListener.onResponse(null);
        }
    }
```

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/transport/RemoteClusterAwareClient.java]: ++ looks good to me. I've pushed the change

**Review by DaveCTurner — APPROVED** [review]
LGTM :)

**Comment by ywelsch** [comment]
Thanks @DaveCTurner !

**Pull request merged**